### PR TITLE
ci: split CI into separate moldy jobs

### PR DESCRIPTION
**Pull request by a10y**
* use [`mold`](https://github.com/rui314/mold) linker for building Rust code
* Splits into several parallel jobs
* disables one test that was taking ~7 minutes with miri, cutting the miri time down considerably

Shaves ~40% off of CI times

### CODE REVIEW

**Comment by a10y** [comment]
Modest improvement, miri still takes the longest here at ~14 minutes. Before splitting out the jobs, the whole build job count on its own take up to 25 minutes. Now because all of the steps are in parallel, the entire non-miri suite completes in ~6.5 minutes

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -15,39 +15,29 @@ env:
...
+    name: "Python (test)"
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: spiraldb/actions/.github/actions/setup-uv@0.2.0
 
+      - name: Python Lint - Format
+        run: uv run ruff format --check .
+      - name: Python Lint - Ruff
+        run: uv run ruff check .
```
Comment (robert3005) [inline .github/workflows/ci.yml]: you don't need to lint here

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -65,6 +55,60 @@ jobs:
...
+    runs-on: ubuntu-latest
+    steps:
+      # setup mold linker
+      - uses: rui314/setup-mold@v1
+      - uses: actions/checkout@v4
+      - uses: ./.github/actions/cleanup
+      - name: Rust Build (Default features)
+        run: cargo build --all-targets
+      - name: Clean cargo to keep disk usage below limit
+        run: cargo clean
```
Comment (robert3005) [inline .github/workflows/ci.yml]: you don't need this?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -65,6 +55,60 @@ jobs:
...
+    name: "Build (all-features)"
+    runs-on: ubuntu-latest
+    steps:
+      - uses: rui314/setup-mold@v1
+      - uses: actions/checkout@v4
+      - uses: ./.github/actions/cleanup
+      - name: Rust Build (All Features)
+        run: cargo build --all-features --all-targets
+      - name: Clean cargo to keep disk usage below limit
+        run: cargo clean
```
Comment (robert3005) [inline .github/workflows/ci.yml]: remove

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -65,6 +55,60 @@ jobs:
...
+      - uses: actions/checkout@v4
+      - name: Docs
+        run: cargo doc --no-deps
+
+  build-default:
+    name: "Build (default)"
+    runs-on: ubuntu-latest
+    steps:
+      # setup mold linker
+      - uses: rui314/setup-mold@v1
```
Comment (robert3005) [inline .github/workflows/ci.yml:69]: don't you need  `- uses: ./.github/actions/setup-rust` somewhere?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -65,6 +55,60 @@ jobs:
           (cd pyvortex && uv run maturin develop)
 
           (cd docs && uv run make html)
+
+  rust-docs:
```
Comment (robert3005) [inline .github/workflows/ci.yml:55]: weirdly the command pyvortex runs also does this

File: .github/workflows/ci.yml (hunk 4 above)
Comment (a10y) [reply .github/workflows/ci.yml:69]: good point. `rustup` is in the default runner image, so I guess it just picks up rust-toolchain file? but better to use the action to get all the other benefits

**Review by robert3005 — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -62,6 +51,59 @@ jobs:
...
+      - name: Rust Build (All Features)
+        run: cargo build --all-features --all-targets
+
+  rust-test:
+    name: "Rust (tests)"
+    runs-on: ubuntu-latest
+    steps:
+      - uses: rui314/setup-mold@v1
+      - uses: actions/checkout@v4
+      - uses: ./.github/actions/cleanup
```
Comment (robert3005) [inline .github/workflows/ci.yml]: You need to run cleanup as the first thing in the job, otherwise it evicts installed binaries

Comment (robert3005) [reply .github/workflows/ci.yml]: But also maybe we don't need it that much if we have separate jobs?

**Comment by robert3005** [comment]
Ok @a10y I think we should change the cleanup job if we are splitting things, it's now taking close to 50% of the time. for some of them now. Could probably trim down the list. I went full hog trying to make sure we don't have to do a cleanup midway through

**Comment by a10y** [comment]
Ok I've removed usage of `cleanup` action from all steps, and have added a new custom action to download a flatc compiler so that we can check all generated files at CI time

**Comment by robert3005** [comment]
looks like you do need some level of cleanup of hosted tools otherwise our builds step runs out of disk

**Comment by a10y** [comment]
It was `node_modules` that was slow to delete. Cleanup times went from 3minutes -> ~10 seconds now

File: .github/workflows/ci.yml (hunk 5 above)
Comment (danking) [reply .github/workflows/ci.yml:55]: yeah `make -C docs html` builds _all_ the docs, including the Rust docs. We could add a different target like `make -C docs full-html` and use that in docs.yml.

**Pull request merged**